### DISCUSSION TITLE
Any way to get pid of browser and chromedriver

### DISCUSSION THREAD

**Discussion by oculunto** [seleniumbase/SeleniumBase#2727] [Q&A] [answered]
Hello, 
I'm trying to get the pid of the browser and the chromedriver processes after opening them with the Driver class.
I can't find a method for it, I can see in the repository that it is used at some point.
I want to be able to force close them if they crash when opening a driver (using Driver class with multiple instances that sometimes overlap).
Thanks again ;)

**Comment by mdmintz** [comment, marked as answer]
For getting process IDs and other system internals, there's a separate library you can use called `psutil`:
https://pypi.org/project/psutil/

But if you're just trying to avoid overlapping drivers in UC Mode during browser launches (when not using `pytest`), then you need to enable SeleniumBase thread-locking (`sys.argv.append("-n")`) so that the browsers don't try to use the same port at the same. I explained how to do that in the 2nd UC Mode video tutorial on YouTube:
https://www.youtube.com/watch?v=2pTpBtaE7SQ&t=1386s

**Reply by oculunto** [reply]
Thank you ! I didn't see that you had a new video I will check that out !
My case is a bit more complicated, I'm running multiple independent instances of a program, and each of them opens browsers as subprocesses too (so far I am using a text file to allow opening or waiting, which does not work quite good), so I don't think they can be joined with ThreadPoolExecutor (or maybe they can ? I am not that advanced).
pytest would require rewriting almost the whole thing

**Reply by mdmintz** [reply]
Maybe the `ThreadPoolExecutor` isn't needed, but the `sys.argv.append("-n")` is definitely needed for the built-in thread-locking to activate.

**Reply by oculunto** [reply]
I will try that out thank you !!

**Reply by oculunto** [reply]
I'm not sure why, but adding it somehow messes with the proxy (I do use other args in my script to launch it with different parameters), but with the -n, all the windows launch with the same proxy (instead of the assigned one)

**Reply by mdmintz** [reply]
Add `multi_proxy=True`. Eg. `SB(multi_proxy=True)` / `Driver(multi_proxy=True)`

https://github.com/seleniumbase/SeleniumBase/blob/b77a1c022bb4ebf8cc1fec4db92847a14f9bc818/seleniumbase/plugins/driver_manager.py#L77

**Reply by oculunto** [reply]
![a8e1a540-93df-48aa-9157-854329f45a10_text](https://github.com/seleniumbase/SeleniumBase/assets/111004839/646432d0-fad3-4e7b-b34f-e9462b712277)
Thank you good sir !

**Reply by oculunto** [reply]
Ok so that worked perfectly.... until ... 
At some point the new generated browsers start using the same proxy (and it ends up blocked with 404).
I investigated a bit and found out that it was due to C:\PYPY\downloaded_files getting filled with proxy_ext_dir_i (where the proxy extensions manifest are created). 
And when it reaches 999 it starts behaving like that.
Is there a way to increase it indefinitely, or at least delete automatically that folder when the browser is closed?

**Reply by mdmintz** [reply]
Use a script like this (with `test=True`) to clean out that folder before it reaches 1000 items:

```python
from seleniumbase import SB

with SB(test=True) as sb:
    pass
```

**Reply by oculunto** [reply]
Thanks for your answer and time.
No way to do it with Driver class instead? To delete it when .quit is invoked ? Or maybe to reuse the same one if data dir provided is the same ? 
Trying to avoid rewriting my whole spaghetti code, I have it all with Driver class. I'll try to patchwork something there, thanks again !

**Reply by mdmintz** [reply]
The `SB()` format has a lot more functionality than the `Driver()` format. `SB()` has access to test framework methods (which include folder cleanup, etc). The `Driver()` format is for people who want a regular Selenium `driver` with some extra functionality, but it doesn't have everything. The `SB()` format also uses a context manager format, which means that it can have special code at the beginning **and at the end** (after leaving the code block).

Since the `Driver()` format doesn't use a context manager format, there is only the special code that runs at the beginning. People are expected to do their own cleanup, eg. `driver.quit()`, and any other cleanup, such as removing the proxy_zip files when no longer needed. In summary, the `Driver()` format gives you some extra flexibility, but at the cost of requiring you to do extra work for some things.

**Reply by oculunto** [reply]
I understand thanks ! I'll try to find a way to clean it up then.
There are more urgent matters now, seems like Cloudflare just caught up, can't pass it anymore (right this minute haha)
EDIT: Nvm it seems like it was a global outage due to maintenance.

**Reply by mdmintz** [reply]
For a moment, I thought I had a repeat situation as https://github.com/seleniumbase/SeleniumBase/issues/2626:
https://www.youtube.com/watch?v=2pTpBtaE7SQ&t=1907s

**Reply by oculunto** [reply]
![HousSX](https://github.com/seleniumbase/SeleniumBase/assets/111004839/ce0cd7e5-68c9-4fc5-b03b-ffe6bfa1286d)

**Comment by OpsecGuy** [comment]
Maybe you wanted this?
```
session = Driver(...)
browser_pid = session.browser_pid
ud_driver_pid = session.service.process.pid
```

**Reply by oculunto** [reply]
Thank you ! Yes if that works it should be exactly what I am looking for (although the error might be coming during the execution of Driver(), so I would never get the Driver object in the first place).
I just want to get rid of both processes if the opening of the Driver fails
